### PR TITLE
[FIX] sale_coupon: promotion program with several identical products

### PR DESCRIPTION
**Pull request by macadelic80**
Issue

    - Create promotion program with product_x
    - Create new quotation
    - Add product_x on the first line
    - Add another product_x on the second line
    - Press "Promotions" button

Cause

    The attribut reward_product_qty except singleton but order_line contain more than one element

Solution

    Take the order line with the higher quantity

opw-2350574

### CODE REVIEW

**Review by nim-odoo — COMMENTED** [review]
@agr-odoo maybe we should take the sum instead of the max?

**Comment by agr-odoo** [comment]
> @agr-odoo maybe we should take the sum instead of the max?

Yes, otherwise in a promotion like 'pay 4 get 2 free' with the same use case we could have 4 lines
|Prod|Qty|
|------|-----|
|Test|    1| 
|Test|    1| 
|Test|    1| 
|Test|    1|
|Test|    1| 
|Test|    1| 

`best_qty` would be `1`
and
`reward_product_qty = min(reward_product_qty, best_qty)` => 1 (wrong, we want 2 free products)

Note that to fully support the case of multiple lines with the same product you need also something like
```diff
diff --git a/addons/sale_coupon/models/sale_order.py b/addons/sale_coupon/models/sale_order.py
index bbaad312d9d..4f1acd6f29c 100644
--- a/addons/sale_coupon/models/sale_order.py
+++ b/addons/sale_coupon/models/sale_order.py
@@ -65,9 +65,9 @@ class SaleOrder(models.Model):
 
     def _is_reward_in_order_lines(self, program):
         self.ensure_one()
-        return self.order_line.filtered(lambda line:
-            line.product_id == program.reward_product_id and
-            line.product_uom_qty >= program.reward_product_quantity)
+        order_quantity = sum(self.order_line.filtered(lambda line:
+            line.product_id == program.reward_product_id).mapped('product_uom_qty'))
+        return order_quantity >= program.reward_product_quantity
 
     def _is_global_discount_already_applied(self):
         applied_programs = self.no_code_promo_program_ids + \

```

**Comment by agr-odoo** [comment]
runbot is still red

**Comment by macadelic80** [comment]
@agr-odoo Done !

File: addons/sale_coupon/models/sale_order.py [hunk 1]
```diff
@@ -65,9 +65,9 @@ def _get_reward_lines(self):
 
     def _is_reward_in_order_lines(self, program):
         self.ensure_one()
-        return self.order_line.filtered(lambda line:
-            line.product_id == program.reward_product_id and
-            line.product_uom_qty >= program.reward_product_quantity)
+        order_quantity = sum(self.order_line.filtered(lambda line:
+         line.product_id == program.reward_product_id).mapped('product_uom_qty'))
```
Comment (nim-odoo) [inline addons/sale_coupon/models/sale_order.py:69]:
```suggestion
            line.product_id == program.reward_product_id).mapped('product_uom_qty'))
```

File: addons/sale_coupon/models/sale_order.py [hunk 2]
```diff
@@ -80,19 +80,20 @@ def _get_reward_values_product(self, program):
 
         order_lines = (self.order_line - self._get_reward_lines()).filtered(lambda x: program._get_valid_products(x.product_id))
         max_product_qty = sum(order_lines.mapped('product_uom_qty')) or 1
+        best_qty = sum(self.order_line.filtered(lambda x: x.product_id == program.reward_product_id).mapped('product_uom_qty'))
```
Comment (nim-odoo) [inline addons/sale_coupon/models/sale_order.py:83]: I think this is more meaningful:
```suggestion
        total_qty = sum(self.order_line.filtered(lambda x: x.product_id == program.reward_product_id).mapped('product_uom_qty'))
```

**Review by nim-odoo — COMMENTED** [review]
robodoo delegate=agr-odoo

File: addons/sale_coupon/models/sale_order.py [hunk 3]
```diff
@@ -65,9 +65,9 @@ def _get_reward_lines(self):
 
     def _is_reward_in_order_lines(self, program):
         self.ensure_one()
-        return self.order_line.filtered(lambda line:
-            line.product_id == program.reward_product_id and
-            line.product_uom_qty >= program.reward_product_quantity)
+        order_quantity = sum(self.order_line.filtered(lambda line:
+          line.product_id == program.reward_product_id).mapped('product_uom_qty'))
```
Comment (nim-odoo) [inline addons/sale_coupon/models/sale_order.py:69]: keep 4 spaces :)

Comment (macadelic80) [reply addons/sale_coupon/models/sale_order.py:69]: Oh ok its 4, sry

**Comment by agr-odoo** [comment]
@robodoo r+

**Pull request closed without merging**